mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+package cat
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/creachadair/taskgroup"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/mempool"
+	"github.com/tendermint/tendermint/proxy"
+	"github.com/tendermint/tendermint/types"
+)
+
+var _ mempool.Mempool = (*TxPool)(nil)
+
+const evictedTxCacheSize = 200
+
+var (
+	ErrTxInMempool       = errors.New("tx already exists in mempool")
+	ErrTxAlreadyRejected = errors.New("tx was previously rejected")
+)
+
+// TxPoolOption sets an optional parameter on the TxPool.
+type TxPoolOption func(*TxPool)
+
+// TxPool implemements the Mempool interface and allows the application to
+// set priority values on transactions in the CheckTx response. When selecting
+// transactions to include in a block, higher-priority transactions are chosen
+// first.  When evicting transactions from the mempool for size constraints,
+// lower-priority transactions are evicted sooner.
+//
+// Within the txpool, transactions are ordered by time of arrival, and are
+// gossiped to the rest of the network based on that order (gossip order does
+// not take priority into account).
+type TxPool struct {
+	// Immutable fields
+	logger       log.Logger
+	config       *config.MempoolConfig
+	proxyAppConn proxy.AppConnMempool
+	metrics      *mempool.Metrics
+
+	updateMtx            sync.Mutex
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // one value sent per height when mempool is not empty
+	preCheckFn           mempool.PreCheckFunc
+	postCheckFn          mempool.PostCheckFunc
+	height               int64 // the latest height passed to Update
+
+	// Thread-safe cache of rejected transactions for quick look-up
+	rejectedTxCache *LRUTxCache
+	// Thread-safe cache of valid txs that were evicted
+	evictedTxs *EvictedTxCache
```

Comment:
Can you please explain the purpose of monitoring evicted transactions?

Reply:
> we can save needing the app to revalidate it.

Thanks for your explanation. Does the evicted tx cache gets reset after each block proposal and approval? asking this as the validity of a tx  is state dependent and may change from one block to another (so the old validity results may not hold for the future state).

Reply:
No txs are only removed from the tx cache if they are committed in a block, if someone sends that same transaction again and the node has adequate space, if `Flush` is called or after a set number of other evicted txs have pushed it out.

`CheckTx` more serves as a very light weight check rather than a comprehensive one based against the current state. The only thing it checks against state for is the validity of the signature (which should not change per height) and the incremental nonce (which can change per height). It's definitely plausible that invalid transactions remain in the mempool until they are committed or removed by a TTL.

However, because they're supposed to be lightweight and fast this means that there isn't much to gain from caching the response anyway